src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,969 @@
+                        tcti_libtpms->state_path,
+                        strerror(errno));
+        }
+    }
+
+    if (tcti_libtpms->state_path != NULL) {
+        free(tcti_libtpms->state_path);
+    }
+
+    if (tcti_libtpms->response_buffer) {
+        free(tcti_libtpms->response_buffer);
+    }
+}
+
+TSS2_RC
+tcti_libtpms_receive(
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *response_size,
+    unsigned char *response_buffer,
+    int32_t timeout)
+{
+#ifdef TEST_FAPI_ASYNC
+    /* Used for simulating a timeout. */
+    static int wait = 0;
+#endif
+
+    TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms = tcti_libtpms_context_cast(tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_libtpms_down_cast(tcti_libtpms);
+    TSS2_RC rc;
+
+    rc = tcti_common_receive_checks(tcti_common, response_size, TCTI_LIBTPMS_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    if (timeout != TSS2_TCTI_TIMEOUT_BLOCK) {
+        LOG_TRACE("Asynchronous I/O not actually implemented.");
+#ifdef TEST_FAPI_ASYNC
+        if (wait < 1) {
+            LOG_TRACE("Simulating Async by requesting another invocation.");
+            wait += 1;
+            return TSS2_TCTI_RC_TRY_AGAIN;
+        } else {
+            LOG_TRACE("Sending the actual result.");
+            wait = 0;
+        }
+#endif /* TEST_FAPI_ASYNC */
+    }
+
+    if (response_buffer == NULL) {
+        *response_size = tcti_libtpms->response_len;
+        return TSS2_RC_SUCCESS;
+    }
+
+    if (*response_size < tcti_libtpms->response_len) {
+        *response_size = tcti_libtpms->response_len;
+        return TSS2_TCTI_RC_INSUFFICIENT_BUFFER;
+    }
+    *response_size = tcti_libtpms->response_len;
+
+    memcpy(response_buffer, tcti_libtpms->response_buffer, tcti_libtpms->response_len);
+
+    LOGBLOB_DEBUG(response_buffer, *response_size, "Response received:");
+
+    free(tcti_libtpms->response_buffer);
+    tcti_libtpms->response_buffer = NULL;
+    tcti_libtpms->response_buffer_len = 0;
+    tcti_libtpms->response_len = 0;
+
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+
+    return TSS2_RC_SUCCESS;
+}
+
+static void
+tcti_libtpms_init_context_data(TSS2_TCTI_COMMON_CONTEXT *tcti_common)
+{
+    TSS2_TCTI_MAGIC (tcti_common) = TCTI_LIBTPMS_MAGIC;
+    TSS2_TCTI_VERSION (tcti_common) = TCTI_VERSION;
+    TSS2_TCTI_TRANSMIT (tcti_common) = tcti_libtpms_transmit;
+    TSS2_TCTI_RECEIVE (tcti_common) = tcti_libtpms_receive;
+    TSS2_TCTI_FINALIZE (tcti_common) = tcti_libtpms_finalize;
+    TSS2_TCTI_CANCEL (tcti_common) = tcti_libtpms_cancel;
+    TSS2_TCTI_GET_POLL_HANDLES (tcti_common) = tcti_libtpms_get_poll_handles;
+    TSS2_TCTI_SET_LOCALITY (tcti_common) = tcti_libtpms_set_locality;
+    TSS2_TCTI_MAKE_STICKY (tcti_common) = tcti_make_sticky_not_implemented;
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+    memset(&tcti_common->header, 0, sizeof(tcti_common->header));
+}
+
+TSS2_RC
+tcti_libtpms_dl(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
```

Comment:
why do we need to this, why wouldn't this just be linked with this at build time?

Reply:
We need to dlopen libtpms because we want to be able to have different libtpms instances. They have no contexts, but a lot of global variables.

Reply:
Got it, interesting way around the problem. For some reason I thought dlopen was RTLD_NOLOAD by default, but I guess not.